Change DrawSprite calls to provide scales instead of sizes.

pchote opened the pull request:
The fundamental motivation behind this PR is that the sprite rendering code has always done its drawing taking arguments for the top-left position and the rendered width/height. This was simple and efficient for our original sprite implementation, and continues to work if you introduce the idea of a virtual frame rectangle *or* pixel depth information. Unfortunately for us, we implement *both* of those things, meaning that things break.

The reason is that sprites that embed a block of pixels inside a larger virtual frame need to define an offset: the margin of empty pixels between the top left corner and the actual start of the pixel data. When the renderer is asked to draw a sprite with a location and size it has to work out how to adjust these offsets to keep the pixels correctly offset within the frame.

The scale is trivial to calculate for x and y because the width and height are never 0: you can divide the requested size by the virtual size and always get a valid scale. However, when we introduce depth, 0 is a valid and common size (meaning that the sprite sits flat in the x-y plane), giving us a z scale factor of 0 / 0. We can't assume that the z scale is the same as the x or y scale, because the code doesn't guarantee that even x and y are the same (and some parts of the UI explicitly depend on them being different).

The current code works around the divide by zero with

https://github.com/OpenRA/OpenRA/blob/dcaa658678c398b600a7b591a4d4cf796ea50493/OpenRA.Game/Graphics/Sprite.cs#L41-L42

simply setting the internal (size-dependent) z offset to 0 if the z size is 0.

This breaks the offset calculation, so we bodge a workaround with

https://github.com/OpenRA/OpenRA/blob/dcaa658678c398b600a7b591a4d4cf796ea50493/OpenRA.Game/Graphics/SpriteRenderable.cs#L81-L82

manually adding the offset to the location that we give to the renderer instead of allowing the renderer to add it internally based on the size. This works fine for sprites that have `ZRamp: 0`, but things start getting messy for bibs, railway tracks, etc that have `ZRamp: 1`. We need to fix this properly before we can correcty solve #12229.

This would have been difficult to fix a few years ago back when the depth code was first implemented, but all the other renderer refactoring since then now makes it simple: we already deal with scale in the high level rendering code, we just need to not multiply it by the sprite size when calling the low level rendering code. This PR does just that, and makes some other tweaks to the DrawSprite overloads to make them more consistent.

There are a few places in the UI (e.g. video rendering) that do need split x and y scale, but these don’t care about depth, so the conversion from size to scale is done at the WidgetUtil level where we can use the type system to guarantee that there are no depth components to blow up.

The changes should hopefully be quite obvious with the explanation above in mind, and I don't expect any changes in the ingame behaviour (except maybe around #12229, but the next PR will fix that properly).

pchote committed:
Rework WidgetUtil sprite rendering helpers.

penev92 commented on OpenRA.Game/Graphics/RgbaSpriteRenderer.cs:
```diff
@@ -22,44 +22,36 @@ public RgbaSpriteRenderer(SpriteRenderer parent)
 		}
 
-		public void DrawSprite(Sprite s, in float3 a, in float3 b, in float3 c, in float3 d)
+		public void DrawSprite(Sprite s, in float3 location, in float scale, in float3 tint, float alpha)
```
Is there a reason the alpha isn't an in parameter?
Same question for the scale on the above method.

pchote replied:
alpha and scale are regular floats, which are too small for the in modifier: https://docs.microsoft.com/en-us/dotnet/csharp/write-safe-efficient-code#use-in-parameters-for-large-structs

abcdefg30 replied:
But `scale` has an in modifier here? 🤔

pchote replied:
oh, thats a typo. Will fix.

pchote committed:
Change DrawSprite calls to provide scales instead of sizes.

This allows us to remove a hacky workaround for calculating
depth offsets when sprites have size.Z == 0.

pchote commented:
Fixed.

reaperrr commented:
https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/5e4617be698baf953cb093c937479e6b31ab8210...b9e263faea652f4fc306a95a154952af0673498a
